### PR TITLE
Relax django/other project dependencies, fix travis

### PR DESCRIPTION
**Pull request by tony**
I believe poetry is used to generate / distribute the sdist package, so we may need to
loosen dependencies in more of these as well.

(Feel free to rebase/fork this PR if I'm not around!)

### CODE REVIEW

**Comment by tony** [comment]
@prokher 9a02dbc should be okay too

https://travis-ci.org/datadvance/DjangoChannelsGraphqlWs/jobs/626191586?utm_medium=notification&utm_source=github_status

> pip._vendor.urllib3.exceptions.ProtocolError: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))

**Comment by tony** [comment]
I can't trigger reruns.

**Comment by tony** [comment]
@prokher Since I've made this PR - I've gotten *some* more experience with poetry.  I reduced the amount of versions changed.

It includes a `poetry update` that updated mypy and other dependencies that made them more strict, so I did some linting fixes to make those things happy.

File: pyproject.toml [hunk 1]
```diff
@@ -47,19 +47,19 @@ packages = [
 
 
 [tool.poetry.dependencies]
-python = "^3.7.0"
+python = "^3.6.0"
 aiohttp = "^3.5"
 asgiref = "^3.2"
 channels = "^2.2"
-django = "^2.2"
```
Comment (tony) [inline pyproject.toml]: This would lock versions out of Django 3: https://github.com/datadvance/DjangoChannelsGraphqlWs/issues/35

File: pyproject.toml [hunk 2]
```diff
@@ -47,19 +47,19 @@ packages = [
 
 
 [tool.poetry.dependencies]
-python = "^3.7.0"
+python = "^3.6.0"
```
Comment (tony) [inline pyproject.toml]: This should reduce the minimum python version required to 3.6.

Even graphql-core-next will only require 3.6: https://github.com/graphql-python/graphql-core-next

So I think this relaxation is safe.

**Comment by tony** [comment]
@prokher @datadvance: Any chance you can give this a check?

**Comment by tony** [comment]
@prokher @datadvance: Hey there! Another ping. I hope things are going okay on your end! If there's anything happening that's getting in the way of updating, that's okay, just let us know!

I think this is safe to go in.  Unless there's any objection you have?

**Comment by tony** [comment]
@prokher @datadvance Hey there!

Are you willing to give me PyPI access and membership on the repo? I can tag the releases if you'd like.  I'm `git-pull` on pypi.

File: .travis.yml [hunk 3]
```diff
@@ -34,8 +34,8 @@ jobs:
           install:
               - pip3 install --upgrade pip
               - pip3 install --progress-bar off poetry
-              - python3 -m poetry --no-interaction install
-              - python3 -m poetry --no-interaction build
+              - python3 -m poetry install --no-interaction
```
Comment (prokher) [inline .travis.yml:37]: Why? What is the purpose of these changes?

File: .travis.yml [hunk 4]
```diff
@@ -46,8 +46,8 @@ jobs:
...
```
Comment (prokher) [inline .travis.yml:49]: Why? What is the purpose of these changes?

File: .travis.yml [hunk 5]
```diff
@@ -61,7 +61,7 @@ jobs:
               - python --version
               - python -m pip install --upgrade pip
               - python -m pip install --progress-bar off poetry
-              - python -m poetry --no-interaction install
-              - python -m poetry --no-interaction build
+              - python -m poetry install --no-interaction
```
Comment (prokher) [inline .travis.yml:64]: Why? What is the purpose of these changes?

File: example/example.py [hunk 6]
```diff
@@ -48,11 +48,14 @@
 # ---------------------------------------------------------------------- TYPES & QUERIES
 
 
-class Message(
```
Comment (prokher) [inline example/example.py:51]: Why? What is the purpose of these changes?

File: pyproject.toml (hunk 2 above)
Comment (prokher) [reply pyproject.toml:50]: Indeed, will do this, but currently problem is not in the requirements, but in the `dataclasses` module usage which appears only in 3.7.

File: pyproject.toml [hunk 7]
```diff
@@ -47,19 +47,19 @@ packages = [
...
-django = "^2.2"
+django = ">=2.2"
 graphene = "^2.1"
 graphql-core = "^2.2"
 msgpack = "^0.6.1"
 
 
 [tool.poetry.dev-dependencies]
 bandit = "*"
-black = { version = "*", allows-prereleases = true }
```
Comment (prokher) [inline pyproject.toml:62]: Again, why? What is the purpose of these changes?

**Comment by tony** [comment]
@prokher Hi, I hope you're doing well and thanks for the review! 😄 

I'm happy addressing any issues. You can see my comments in individual commits for more information if you'd like. 

You are welcome to fork my branch, you are also welcome to get Travis working and django compatible with 3.x in your own way if you prefer!

I will respond to inline, my purpose is to:

- Fix errors that were coming up on Travis (master is currently failing)
- Fix #35. when using pipenv, this package is limited to django 2.x. Locking will fail and means projects pulling in this package are locked into django 2.x until the dependency version constraint is relaxed. Does that make sense?

I originally written this when I was newer to Poetry. I also wanted to get tests passing on the project.

File: pyproject.toml (hunk 7 above)
Comment (tony) [reply pyproject.toml:62]: This was found inside travis.

Poetry 1.0 changes the field name: https://github.com/python-poetry/poetry/releases/tag/1.0.0

> Specifying dependencies with `allows-prereleases` in the pyproject.toml file is deprecated for consistency with the add command. Use `allow-prereleases` instead.

Newer poetry versions renamed `allow-prereleases` to `allow-prereleases`. This caused the travis build to fail.

It also updates black.

https://github.com/datadvance/DjangoChannelsGraphqlWs/pull/37/commits/7a772583d3421f89f978ea0639ae688c50e549d3

image

File: example/example.py (hunk 6 above)
Comment (tony) [reply example/example.py:51]: This, (and the other ones similar to this) were to fix mypy on Travis.

https://travis-ci.org/github/datadvance/DjangoChannelsGraphqlWs/jobs/630643421?utm_medium=notification&utm_source=github_status

```
>                   stderr)
E               plumbum.commands.processes.ProcessExecutionError: Unexpected exit code: 1
E               Command line: | /home/travis/virtualenv/python3.7.1/bin/mypy example/
E               Stdout:       | example/example.py:51: error: Unexpected keyword argument "default_resolver" for "__init_subclass__" of "object"
```

image


(Also, pardon me!): I'm just not familiar with the use of classes like this.  What is `default_resolver=X` doing from a python perspective?  Being passed along into `__init_subclass__`? Is there a place in the python documentation for how this works or what it's called?

File: pyproject.toml (hunk 7 above)
Comment (tony) [reply pyproject.toml:62]: https://travis-ci.org/github/datadvance/DjangoChannelsGraphqlWs/jobs/626173702?utm_medium=notification&utm_source=github_status

image

```
The "black" dependency specifies the "allows-prereleases" property, which is deprecated. Use "allow-prereleases" instead.
ImportError while loading conftest '/home/travis/build/datadvance/DjangoChannelsGraphqlWs/tests/conftest.py'.
tests/conftest.py:27: in <module>
    import channels
E   ModuleNotFoundError: No module named 'channels'
```

(Thankfully travis still has these!)

File: .travis.yml (hunk 3 above)
Comment (tony) [reply .travis.yml:37]: Poetry 1.0 wants flags (e.g. `--no-interaction`) to be after the subcommand

https://travis-ci.org/github/datadvance/DjangoChannelsGraphqlWs/jobs/626173702#L404

image

File: .travis.yml (hunk 4 above)
Comment (tony) [reply .travis.yml:49]: Same as https://github.com/datadvance/DjangoChannelsGraphqlWs/pull/37/commits/288f9e75e34a6d3f38d9eda10ee8b39b80125741#r403520450

File: .travis.yml (hunk 5 above)
Comment (tony) [reply .travis.yml:64]: See https://github.com/datadvance/DjangoChannelsGraphqlWs/pull/37/commits/288f9e75e34a6d3f38d9eda10ee8b39b80125741#r403520450

**Comment by tony** [comment]
@prokher Hi I wanted to come back and say - I could have done more to point out that stuff in the description. Poetry v1.0 changes some things, and mypy was failing on travis. In addition pipenv and other package managers get the django dependency locked at 2.x (#35)

I also notice on master things are switching to GH actions? The reason I ask is this PR fixes the CI for travis.

**Comment by prokher** [comment]
@tony I am just playing with GH actions. Actually there is already GitLab pipelines which run tests, but this one runs inside the company perimeter. I am now trying to fix Travis CI. So I would like it to work as well, at least until it takes too much time to maintain - Travis does strange things time after time, especially on Windows ;-)

BTW, I realized why you changed Poetry command line parameters, `--no-interaction` is a "global" parameter belonging to the Poetry itself, but not to the `install` subcommand. Although both variants will work, I will switch to the one you recommended in this PR.

**Comment by prokher** [comment]
> BTW, I realized why you changed Poetry command line parameters, --no-interaction is a "global" parameter belonging to the Poetry itself, but not to the install subcommand. Although both variants will work, I will switch to the one you recommended in this PR.

or not 🤦‍♂

**Comment by tony** [comment]
You are welcome to cherry pick my commits from this PR

Also I can rebase off master if youd like!

**Comment by prokher** [comment]
@tony, please take a look, is there anything useful left in this PR, please tell me if so. At the moment, I thing everything done here is already covered in the `master` branch.

**Comment by tony** [comment]
@prokher From here I think I will close (it is getting a bit later by me, 7PM and I'm tired, but it's 3AM by you in Moscow?)

Thank you for updating the package!  I will make a new PR tomorrow if there's anything left to add / look at.

**Comment by prokher** [comment]
@tony Thank you very much. Indeed it was a deep night in the Moscow. It often happens like this when I do a small fix before going to sleep 🤦‍♂ Anyway, I think it was worth doing eventually. I own the community (and you, in particular) these fixes at least since the beginning of 2020, so I am happy I eventually went back to this project. Have a nice weekend.

**Comment by tony** [comment]
@prokher Have a good weekend!  Hope things are going well over there!
